Shell: printf available safely from different threads

jakub-uC opened the pull request:
Shell should managed correct text display when function shell_fprintf and print macros will be used from different threads.

Commit:
shell: removed foreground command functionality

Removed foreground command functionality from shell source files.
Removed associated example.
Removed enter/exit command functions from the Bluetooth example
Updated project config files.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

carlescufi commented:
Thanks @jarz-nordic. I have marked this as DNM until it's ready, but this is much appreciated, especially from the Bluetooth side!

carlescufi commented on subsys/shell/shell.c:
```diff
@@ -1072,22 +1097,20 @@ static void shell_log_process(const struct shell *shell)
+		 * readable and can be used to enter further commands.
+		 */
+		if (shell->ctx->cmd_buff_len) {
+			k_sleep(K_MSEC(15));
```
Should this be made configurable? Or even better, could you use perhaps `k_yield()` instead?

jarz-nordic replied:
It is not related to this change. @nordic-krch  what is your opinion about it?

carlescufi replied:
No need to move to Kconfig now if we haven't seen any issues yet with it not being configurable. Thanks for the explanation.

dbkinder commented on doc/subsystems/shell/shell.rst:
```diff
@@ -241,39 +243,9 @@ Simple command handler implementation:
-		/* Terminate foreground command. */
-		shell_command_exit(foo_shell);
-	}
+:c:macro:`shell_error` can be used from the command handler or from threads, it
```
This would read better as:
```
... from the command handler or from threads, but not
from an interrupt context.  Instead, interrupt handlers
should use :ref:`logger` for printing.
```

jarz-nordic replied:
done

nordic-krch commented on subsys/shell/shell.c:
```diff
@@ -28,6 +28,31 @@
+
+	va_start(args, p_fmt);
+
+	if (IS_ENABLED(CONFIG_SHELL_VT100_COLORS) &&
```
this if-else statement is the same for `shell_fprintf`. Could you extract it to separate function?

jarz-nordic replied:
done

nordic-krch commented on subsys/shell/shell_ops.c:
```diff
@@ -412,8 +415,11 @@ void shell_vt100_color_set(const struct shell *shell,
 {
-	shell_vt100_color_set(shell, color->col);
-	vt100_bgcolor_set(shell, color->bgcol);
+	if (IS_ENABLED(CONFIG_SHELL_VT100_COLORS)) {
```
i think that it's not needed. Calls to this function (and `shell_vt100_colors_store`) in shell.c are surrounded by `if(CONFIG_SHELL_VT100_COLORS)` so if colors are off linker will remove those functions as unused.

jarz-nordic commented:
@carlescufi @nordic-krch @rlubos 
I think it is ready

Commit:
shell: safe print from different threads

Added display text management to shell_fprintf function.
Now it can be used from diffrent threads with not risk that
displayed lines will overlay.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>
Signed-off-by: Krzysztof Chruściński <krzysztof.chruscinski@nordicsemi.no>

Commit:
shell: documentation update

Updated documantation in regards to new shell printing strategy.

Signed-off-by: Jakub Rzeszutko <jakub.rzeszutko@nordicsemi.no>

rlubos reviewed (approved):
Works fine with OpenThread 👍

nordic-krch commented:
@jarz-nordic looks very good. one minor comment and i'll approve.

jarz-nordic commented:
@nordic-krch : done
